File: 2022/day2.nim
include prelude

var 
  s1 = 0
  s2 = 0

for line in "day2.txt".lines:
  let
    elf = int(line[0]) - int('A') + 1
    me = int(line[2]) - int('X') + 1
  s1 += me
  if elf mod 3 + 1 == me:
    s1 += 6
  elif elf == me:
    s1 += 3
  else:
    s1 += 0
  s2 += me * 3 - 3
  if me == 3: s2 += elf mod 3 + 1
  elif me == 2: s2 += elf
  elif elf != 1: s2 += (elf + 2) mod 3
  else: s2 += 3

echo &"{s1} {s2}"

File: python/advent2024/day2.py
from itertools import pairwise
from pathlib import Path

lines = Path("day2.txt").read_text().splitlines()

reports = {tuple(map(int, line.split())) for line in lines}
# ruff: noqa: PLR2004
safe_reports = {
  report
  for report in reports
  if (
    all(a <= b and 1 <= abs(a - b) <= 3 for a, b in pairwise(report))
    or all(a >= b and 1 <= abs(a - b) <= 3 for a, b in pairwise(report))
  )
}
safe_ish_reports = safe_reports | {
  report
  for report in reports.difference(safe_reports)
  if any(
    all(a <= b and 1 <= abs(a - b) <= 3 for a, b in pairwise(redacted))
    or all(a >= b and 1 <= abs(a - b) <= 3 for a, b in pairwise(redacted))
    for redacted in ((*report[:i], *report[i + 1 :]) for i in range(len(report)))
  )
}

print(
  len(safe_reports),
  len(safe_ish_reports),
)
